src/webargs/multidictproxy.py

```diff
@@ -0,0 +1,67 @@
+from webargs.compat import MARSHMALLOW_VERSION_INFO, Mapping
+from webargs.core import missing, is_multiple
+
+
+class MultiDictProxy(Mapping):
+    """
+    A proxy object which wraps multidict types along with a matching schema
+    Whenever a value is looked up, it is checked against the schema to see if
+    there is a matching field where `is_multiple` is True. If there is, then
+    the data should be loaded as a list or tuple.
+
+    In all other cases, __getitem__ proxies directly to the input multidict.
+    """
+
+    def __init__(self, multidict, schema):
+        self.data = multidict
+        self.multiple_keys = self._collect_multiple_keys(schema)
+
+    def _collect_multiple_keys(self, schema):
+        result = set()
+        for name, field in schema.fields.items():
+            if not is_multiple(field):
+                continue
+            if MARSHMALLOW_VERSION_INFO[0] < 3:
+                result.add(field.load_from if field.load_from is not None else name)
+            else:
+                result.add(field.data_key if field.data_key is not None else name)
+        return result
+
+    def __getitem__(self, key):
+        val = self.data.get(key, missing)
+        if val is not missing and key in self.multiple_keys:
+            if hasattr(self.data, "getlist"):
+                return self.data.getlist(key)
+            elif hasattr(self.data, "getall"):
+                return self.data.getall(key)
+            elif isinstance(val, (list, tuple)):
+                return val
+            if val is None:
+                return None
+            else:
+                return [val]
+        return val
+
+    def __delitem__(self, key):
+        del self.data[key]
+
+    def __setitem__(self, key, value):
```

Comment:
I don't think so, because it will fail with an appropriate error the way that this is now. Replicating the work in the frameworks to make these immutable just seems to offer little value to me.

If a Flask user ended up handling the `MultiDictProxy` (e.g. a validator which takes original data), using `__setitem__` will give a `TypeError` because it just calls `ImmutableMultiDict.__setitem__` anyway. That seems like pretty good behavior to me.

Is there a case in which this will fall over?

I am looking at those classes and wondering if we should proxy the `copy` method for Flask though. I wouldn't be surprised if other frameworks have useful specialty-methods which people might want to use too.

Reply:
Fair point; I think this is fine as is. I'm not sure about `copy`; I don't know when that would come up in practice.

Reply:
Maybe the best option for `copy` and its ilk is to ignore it. And if someone comes along requesting that `MultiDictProxy` also proxies `copy`, it can be added pretty trivially.

Reply:
Sounds good